src/App.jsx

```diff
@@ -6,14 +6,36 @@ function App() {
   const [monthlyContribution, setMonthlyContribution] = useState('')
   const [annualReturn, setAnnualReturn] = useState('7')
   const [yearsToRetirement, setYearsToRetirement] = useState('')
+  
+  // Real rate of return calculation inputs
+  const [nominalReturn, setNominalReturn] = useState('')
+  const [inflationRate, setInflationRate] = useState('')
+  
+  // Social security inputs
+  const [socialSecurityMonthly, setSocialSecurityMonthly] = useState('')
+  const [yearsUntilSocialSecurity, setYearsUntilSocialSecurity] = useState('')
+  
   const [results, setResults] = useState(null)
 
+  const calculateRealReturn = () => {
+    const nominal = parseFloat(nominalReturn) / 100 || 0
+    const inflation = parseFloat(inflationRate) / 100 || 0
+    
+    // Real rate = (1 + nominal) / (1 + inflation) - 1
```

Comment:
⚖️ **Midlevel Developer**

Just a heads-up: In the `calculateRealReturn` function (see line 17-26 in `src/App.jsx`), the formula for the real rate of return is correct, but you might want to guard against negative inflation rates or zero values to avoid weird results. At my last company, we always included input validation here, and sometimes even unit tests for edge cases like hyperinflation or deflation (yes, it happened). Consider adding some simple checks and maybe a couple Jest tests for this function to ensure reliability!